prow/owners/owners.go

```diff
@@ -0,0 +1,477 @@
+		owners = entry.owners.filterCollaborators(collaborators)
+	}
+	return owners, nil
+}
+
+func mdYAMLEnabledFromConfig(ca *config.Agent) func(org, repo string) bool {
+	return func(org, repo string) bool {
+		enabledRepos := ca.Config().Owners.MDYAMLRepos
+		full := fmt.Sprintf("%s/%s", org, repo)
+		for _, elem := range enabledRepos {
+			if elem == org || elem == full {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+func (a RepoAliases) ExpandAlias(alias string) sets.String {
+	if a == nil {
+		return nil
+	}
+	return a[github.NormLogin(alias)]
+}
+
+func (a RepoAliases) ExpandAliases(logins sets.String) sets.String {
+	if a == nil {
+		return logins
+	}
+	// Make logins a copy of the original set to avoid modifying the original.
+	logins = logins.Union(nil)
+	for _, login := range logins.List() {
+		if expanded := a.ExpandAlias(login); len(expanded) > 0 {
+			logins.Delete(login)
+			logins = logins.Union(expanded)
+		}
+	}
+	return logins
+}
+
+func loadAliasesFrom(baseDir string, log *logrus.Entry) RepoAliases {
+	path := filepath.Join(baseDir, aliasesFileName)
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to read alias file %q. Using empty alias map.", path)
+		return nil
+	}
+	config := &struct {
+		Data map[string][]string `json:"aliases,omitempty"`
+	}{}
+	if err := yaml.Unmarshal(b, config); err != nil {
+		log.WithError(err).Errorf("Failed to unmarshal aliases from %q. Using empty alias map.", path)
+		return nil
+	}
+
+	result := make(RepoAliases)
+	for alias, expanded := range config.Data {
+		result[github.NormLogin(alias)] = normLogins(expanded)
+	}
+	log.Infof("Loaded %d aliases from %q.", len(result), path)
+	return result
+}
+
+func loadOwnersFrom(baseDir string, mdYaml bool, aliases RepoAliases, log *logrus.Entry) (*RepoOwners, error) {
+	o := &RepoOwners{
+		RepoAliases:  aliases,
+		baseDir:      baseDir,
+		enableMDYAML: mdYaml,
+		log:          log,
+
+		approvers: make(map[string]sets.String),
+		reviewers: make(map[string]sets.String),
+		labels:    make(map[string]sets.String),
+	}
+
+	return o, filepath.Walk(o.baseDir, o.walkFunc)
+}
+
+// by default, github's api doesn't root the project directory at "/" and instead uses the empty string for the base dir
+// of the project. And the built-in dir function returns "." for empty strings, so for consistency, we use this
+// canonicalize to get the directories of files in a consistent format with NO "/" at the root (a/b/c/ -> a/b/c)
+func canonicalize(path string) string {
+	if path == "." {
+		return baseDirConvention
+	}
+	return strings.TrimSuffix(path, "/")
+}
+
+func (o *RepoOwners) walkFunc(path string, info os.FileInfo, err error) error {
+	log := o.log.WithField("path", path)
+	if err != nil {
+		log.WithError(err).Error("Error while walking OWNERS files.")
+		return nil
+	}
+	filename := filepath.Base(path)
+	if info.Mode().IsDir() {
+		switch filename {
+		case ".git":
+			return filepath.SkipDir
+		case "_output":
```

Comment:
Do we want to have a whiteliest of these?

Reply:
I changed this to use a hard-coded set of directory names as a blacklist. It can be changed to be parameterized as a configuration option.

Maybe the `vendor` or `staging` directories should be included in the blacklist to prevent vendored OWNERS files from having an effect? Currently the approval-handler uses any OWNERS files it finds in these directories. I'm not sure what the desired behavior is for those directories.